r/R/main.R

```diff
@@ -153,3 +125,26 @@ calc_Ks <- function(ks = NULL, temp_c = 25, sal = 35, p_bar = NULL, magnesium =
 
   return(ks_value)
 }
+
+#' @title Calculate equilibrium constants for seawater
+#'
+#' @description Wrapper to calculate all stoichiometric equilibrium constants at given temperature, salinity, pressure and the concentration of magnesium, calcium, sulphate, and fluorine.
+#'
+#' @author Ross Whiteford
+#'
+#' @param temp_c Temperature (Celsius)
+#' @param p_bar Pressure (Bar) (optional)
+#' @param sal Salinity
+#' @param magnesium Magnesium concentration in mol/kgsw. If None, modern is assumed (0.0528171). Should be the average magnesium concentration in seawater - a salinity correction is then applied to calculate the magnesium concentration in the sample.
+#' @param calcium Calcium concentration in mol/kgsw. If None, modern is assumed (0.0102821). Should be the average calcium concentration in seawater - a salinity correction is then applied to calculate the magnesium concentration in the sample.
+#' @param sulphate Sulphate concentration in mol/kgsw. Calculated from salinity if not given.
+#' @param fluorine Fluorine concentration in mol/kgsw. Calculated from salinity if not given.
+#' @param method Options: `R_Polynomial`, `MyAMI_Polynomial` , `MyAMI` (defaults to "MyAMI").
+#' @return Data.table of specified Ks at the given conditions
+#' @export
+calc_all_Ks <- function(temp_c = 25, sal = 35, p_bar = NULL, magnesium = 0.0528171, calcium = 0.0102821, sulphate = NULL, fluorine = NULL, method = "MyAMI") {
```

Comment:
I don't understand why we need an extra wrapper?

Reply:
This is a little finicky and it's all Matlab's fault.
R and Python had the two functions (I'm simplifying a bit here but it was close to):
`calc_K(k, temp_c, sal, p_bar, magnesium, calcium)`
`calc_Ks(ks, temp_c, sal, p_bar, magnesium, calcium)`.
In R and Python `ks` was sensibly made into an optional input with default `ks=None`, and then if `ks` was None it would get all the K*'s.
There's no such thing as default input arguments in Matlab, there's no such thing as keyword arguments, and there's no such thing as `None`. It's all about the position of the input argument (e.g. is it the first variable to go in, or the fifth, or the tenth).  Maintaining the order of the input arguments is critical for Matlab. So I figured I could achieve that by having the two functions as above (but `ks` is mandatory in constrast to before), then a third function which can omit the `ks` variable. Then they have nearly identical function signatures.

This does raise one point though which is that temperature and salinity are optional inputs in both R and Python whereas Matlab they are necessary. While that could be worked around it would be such ugly syntax....because it's positional. So where in python you could write: `kgen.calc_all_Ks(p_bar=30)` in Matlab you'd have to write `kgen.calc_all_Ks(NaN,NaN,30)`.

Reply:
I guess it depends what you consider user friendliness - one function that does everything or multiple functions that each do one thing.
I can see why having a single function called `calc_Ks` might sometimes be more user friendly.
I'm not convinced having three functions `calc_K`, `calc_Ks`, and `calc_all_Ks` is any less clear. To me this is more explicit.

One of the selling points of Kgen is the consistent interface between languages, otherwise we could just have programmed all three implementations separately (in an idiosyncratic way for each languages) and done a comparison.